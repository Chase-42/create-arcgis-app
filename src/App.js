import React from 'react';
import { Route } from 'react-router-dom';
import { UserSession } from '@esri/arcgis-rest-auth';
import './App.scss';
import AppNav from './components/AppNav';
import UserMenu from './components/UserMenu';
import Home from './routes/Home';
import Items from './routes/Items';

function getSessionFromLocalStorage () {
  const serializedSession = window && window.localStorage && window.localStorage.getItem('CAA_SESSION');
  const session = serializedSession && UserSession.deserialize(serializedSession);
  return session;
}

// save session & user for next time the user loads the app
function saveSessionToLocalStorage (session) {
  if (!session || !window || !window.localStorage) {
    return;
  }
  window.localStorage.setItem('CAA_SESSION', session.serialize());
}

function removeSessionFromLocalStorage() {
  if (!window || !window.localStorage) {
    return;
  }
  window.localStorage.removeItem('CAA_SESSION');
}

class App extends React.Component {
  constructor(props) {
    super(props)

    // initialize app from previous session (if any)
    const session = getSessionFromLocalStorage();
    this.state = {
      session
    };
  }
  signIn = () => {
    // sign in
    UserSession.beginOAuth2({
      // TODO: get clientId from config
      clientId: 'EICkmTOXkBhPwIRp',
      popup: true,
      redirectUri: `${window.location.origin}/redirect.html`
    })
    .then(session => {
      // make session available to the app
      this.setState({ session });
      // save session for next time the user loads the app
      saveSessionToLocalStorage(session);
      // get user info
      this.initUser(session);
    });
  }
  initUser = (session) => {
    // fetch user info
    session.getUser()
    .then(currentUser => {
      // make user available to the app
      this.setState({ currentUser });
    });
  }
  signOut = () => {
    // signal to app that the user has signed out
    this.setState({
      session: null,
      currentUser: null
    });
    // make sure the user is not logged in the next time they load the app
    removeSessionFromLocalStorage();
  }
  componentDidMount () {
    const {
      session
    } = this.state;
    if (session) {
      // we have a previously saved session, get the user info
      this.initUser(session);
    }
  }
  render () {
    const { currentUser } = this.state;
    // NOTE: we bind the user menu and render it here 
    // and pass it to the nav menu in order to avoid prop drilling
    // see: https://reactjs.org/docs/context.html#before-you-use-context
    const userMenu =  (<UserMenu currentUser={currentUser} onSignIn={this.signIn} onSignOut={this.signOut} />);
    return (
      <>
        <AppNav userMenu={userMenu} />
        <div className="container mt-5">
          <Route exact path="/" component={Home} />
          <Route path="/items" component={Items} />
        </div>
      </>
    );  
  }
}

export default App;
