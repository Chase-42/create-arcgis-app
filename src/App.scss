// Import Bootstrap and its default variables
@import '~bootstrap/scss/bootstrap.scss';

// global styles
body {
  padding-top: 3.5rem;
}
