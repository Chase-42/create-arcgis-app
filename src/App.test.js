import React from 'react';
import { render, wait } from 'react-testing-library';
import { BrowserRouter as Router } from 'react-router-dom';
import App from './App';

describe('smoke tests', function() {
  describe('w/o a previous session', function() {
    const title = 'Test Title';
    it('renders app title and sign in button', () => {
      const { getByText } = render(
        <Router>
          <App title={title} />
        </Router>
      );
      expect(getByText(title)).toBeInTheDocument();
      expect(getByText('Sign In')).toBeInTheDocument();
    });
    it('matches snapshot', () => {
      const { asFragment } = render(
        <Router>
          <App title={title} />
        </Router>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
