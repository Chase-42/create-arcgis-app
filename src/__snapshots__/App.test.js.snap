// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke tests w/o a previous session matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-md navbar-dark bg-dark fixed-top"
  >
    <a
      class="navbar-brand"
      href="#"
    >
      Test Title
    </a>
    <button
      class="navbar-toggler"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
    >
      <ul
        class="navbar-nav"
      >
        <li
          class="nav-item"
        >
          <a
            aria-current="page"
            class="nav-link active"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/items"
          >
            Items
          </a>
        </li>
      </ul>
      <ul
        class="ml-auto navbar-nav"
      >
        <li
          class="ml-auto nav-item"
        >
          <button
            class="nav-link btn btn-link"
            type="button"
          >
            Sign In
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    class="container mt-5"
  >
    <div
      class="jumbotron"
    >
      <h1
        class="display-3 text-light text-center mb-5"
      >
        Test Title
      </h1>
      <form
        class="search-form undefined"
      >
        <div
          class="input-group input-group-lg"
        >
          <input
            class="form-control"
            placeholder="search for items"
            value=""
          />
          <div
            class="input-group-append"
          >
            <button
              class="btn btn-secondary"
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
